File: src/ros2_bridge/mappers/nav2_msgs/missed_waypoint.rs
//! Typed mapper for `nav2_msgs/msg/MissedWaypoint`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn missed_waypoint_to_bus(msg: ros_env::nav2_msgs::msg::MissedWaypoint) -> crate::nav2_msgs::msg::v1::MissedWaypoint {
    crate::nav2_msgs::msg::v1::MissedWaypoint {
        index: msg.index,
        goal: Some(crate::ros2_bridge::mappers::geometry_msgs::pose_stamped::pose_stamped_to_bus(msg.goal)),
        error_code: msg.error_code,
    }
}

pub(crate) fn missed_waypoint_to_ros(bus: crate::nav2_msgs::msg::v1::MissedWaypoint) -> ros_env::nav2_msgs::msg::MissedWaypoint {
    ros_env::nav2_msgs::msg::MissedWaypoint {
        index: bus.index,
        goal: crate::ros2_bridge::mappers::geometry_msgs::pose_stamped::pose_stamped_to_ros(bus.goal.unwrap_or_default()),
        error_code: bus.error_code,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct Nav2MsgsMissedWaypointMapper;

impl TypedTopicMapper for Nav2MsgsMissedWaypointMapper {
    type Ros = ros_env::nav2_msgs::msg::MissedWaypoint;
    type Bus = crate::nav2_msgs::msg::v1::MissedWaypoint;

    fn type_name(&self) -> &'static str {
        "nav2_msgs/msg/MissedWaypoint"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(missed_waypoint_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(missed_waypoint_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav2_msgs/missed_waypoint.py
"""Generated mapper for `nav2_msgs/msg/MissedWaypoint`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose_stamped import pose_stamped_to_bus, pose_stamped_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.nav2_msgs.msg.v1 import MissedWaypoint as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def missed_waypoint_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.index = msg.index
    bus.goal.CopyFrom(pose_stamped_to_bus(msg.goal))
    bus.error_code = msg.error_code
    return bus


def missed_waypoint_to_ros(bus):
    from nav2_msgs.msg import MissedWaypoint as RosMsg

    out = RosMsg()
    out.index = bus.index
    out.goal = pose_stamped_to_ros(bus.goal)
    out.error_code = bus.error_code
    return out


class Nav2MsgsMissedWaypointMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from nav2_msgs.msg import MissedWaypoint as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return missed_waypoint_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return missed_waypoint_to_ros(bus)
